server/src/main/java/io/crate/analyze/TableElementsAnalyzer2.java

```diff
@@ -0,0 +1,596 @@
+                            "INDEX constraint cannot be used on columns of type \"%s\"", builder.type));
+                    }
+                } else if (constraint instanceof NotNullColumnConstraint<Expression> notNull) {
+                    builder.nullable = false;
+                } else if (constraint instanceof PrimaryKeyColumnConstraint<Expression> primaryKey) {
+                    markAsPrimaryKey(builder);
+                }
+            }
+
+            ColumnType<Expression> type = columnDefinition.type();
+            while (type instanceof CollectionColumnType collectionColumnType) {
+                type = collectionColumnType.innerType();
+            }
+            if (type instanceof ObjectColumnType<Expression> objectColumnType) {
+                builder.columnPolicy = objectColumnType.columnPolicy().orElse(ColumnPolicy.DYNAMIC);
+                for (ColumnDefinition<Expression> nestedColumn : objectColumnType.nestedColumns()) {
+                    nestedColumn.accept(this, columnName);
+                }
+            }
+
+            return null;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Void visitPrimaryKeyConstraint(PrimaryKeyConstraint<?> node, ColumnIdent parent) {
+            PrimaryKeyConstraint<Expression> pkConstraint = (PrimaryKeyConstraint<Expression>) node;
+            List<Expression> pkColumns = pkConstraint.columns();
+
+            for (Expression pk : pkColumns) {
+                Symbol pkColumn = toSymbol.apply(pk);
+                ColumnIdent columnIdent = Symbols.pathFromSymbol(pkColumn);
+                RefBuilder column = columns.get(columnIdent);
+                if (column == null) {
+                    throw new ColumnUnknownException(columnIdent, table);
+                }
+                markAsPrimaryKey(column);
+            }
+
+            return null;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Void visitIndexDefinition(IndexDefinition<?> node, ColumnIdent parent) {
+            IndexDefinition<Expression> indexDefinition = (IndexDefinition<Expression>) node;
+            String name = indexDefinition.ident();
+            ColumnIdent columnIdent = parent == null ? new ColumnIdent(name) : ColumnIdent.getChildSafe(parent, name);
+            RefBuilder builder = columns.get(columnIdent);
+            builder.indexMethod = indexDefinition.method();
+            builder.indexProperties = indexDefinition.properties().map(toSymbol);
+            builder.indexSources = Lists2.map(indexDefinition.columns(), toSymbol);
+            builder.indexType = switch (builder.indexMethod.toLowerCase(Locale.ENGLISH)) {
+                case "fulltext" -> IndexType.FULLTEXT;
+                case "off" -> IndexType.NONE;
+                case "plain" -> IndexType.PLAIN;
+                default -> IndexType.PLAIN;
+            };
```

Comment:
Maybe worth moving this into a `IndexType.of()` method (we already have 2 occurrences of this)?